### PR TITLE
SLVS-960 CFG creation should handle "null" as a case label

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ControlFlowGraphBuilder.cs#L329](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/0de2028cf977831f224b6da55556419871b33d99/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L329): Either merge this case with the identical one on line "292" or change one of the implementations. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1871)
1. MAJOR [ControlFlowGraphBuilder.cs#L333](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/0de2028cf977831f224b6da55556419871b33d99/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L333): Either merge this case with the identical one on line "292" or change one of the implementations. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1871)
1. MAJOR [ControlFlowGraphBuilder.cs#L449](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/0de2028cf977831f224b6da55556419871b33d99/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L449): Either merge this case with the identical one on line "270" or change one of the implementations. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1871)
1. INFO [ControlFlowGraphBuilder.cs#L669](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/0de2028cf977831f224b6da55556419871b33d99/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L669): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -513,12 +514,19 @@ private void BuildGotoCaseStatement(GotoStatementSyntax statement)
             }
 
             var currentJumpBlocks = SwitchGotoJumpBlocks.Peek();
-            if (!currentJumpBlocks.ContainsKey(constValue.Value))
+
+            var indexer = constValue.Value;
+            if (indexer == null)
```
Comment (m-g-sonar) [inline src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: The `null` value of the expression is evaluated and directly accessible as `null` ?

File: src/Tests/SonarLint.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs [hunk 2]
```diff
@@ -1520,7 +1520,7 @@ public void Cfg_Switch_NoDefault()
         [TestCategory("CFG")]
         public void Cfg_Switch_GotoCase()
         {
-            var cfg = Build("cw0(); switch(a) { case 1: case 2: cw1(); goto case 3; case 3: default: case 4: cw2(); break; } cw3();");
+            var cfg = Build("cw0(); switch(a) { case 1: case 2: cw1(); goto case null; case null: default: case 4: cw2(); break; } cw3();");
```
Comment (m-g-sonar) [inline src/Tests/SonarLint.UnitTest/Helpers/FlowAnalysis/ControlFlowGraphTest.cs]: You removed the double `case 3`,  in favor of 2 `case null`,  which means that you are now testing two different things at the same time. I would definitely keep the double `3` and add a single `null` in order to explicitly cover the 2 situations.

File: src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs [hunk 3]
```diff
@@ -588,7 +596,13 @@ private void BuildSwitchStatement(SwitchStatementSyntax switchStatement)
                         {
                             throw new InvalidOperationException("Expression has no constant value");
                         }
-                        caseBlocksByValue[constValue.Value] = fallThroughBlock;
+
+                        var indexer = constValue.Value;
```
Comment (m-g-sonar) [inline src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: You are doing the same thing two times. Can't you extract a method retrieving a value from an expression directly? It's the same behavior to get the `GotoNullEntry` objects for `SwitchStatementSyntax` and `GotoCaseSyntax` nodes.

File: src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs (hunk 1 above)
Comment (tvsonar) [inline src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs]: Yes, `constValue` is an `Optional<object>` (note, not `Nullable<>`). When the constant it represents is `null`, then `constValue.Value` is `null`.

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ControlFlowGraphBuilder.cs#L329](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/4b21fb4b41732ffdd8e0ee6648be249dc962820e/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L329): Either merge this case with the identical one on line "292" or change one of the implementations. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1871)
1. MAJOR [ControlFlowGraphBuilder.cs#L333](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/4b21fb4b41732ffdd8e0ee6648be249dc962820e/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L333): Either merge this case with the identical one on line "292" or change one of the implementations. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1871)
1. MAJOR [ControlFlowGraphBuilder.cs#L449](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/4b21fb4b41732ffdd8e0ee6648be249dc962820e/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L449): Either merge this case with the identical one on line "270" or change one of the implementations. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1871)
1. INFO [ControlFlowGraphBuilder.cs#L665](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/4b21fb4b41732ffdd8e0ee6648be249dc962820e/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L665): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. INFO [ControlFlowGraphBuilder.cs#L659](https://github.com/SonarSource-VisualStudio/sonaranalyzer-csharp/blob/f89123f8275e1ca9da22d74200f2bdba16031afc/src/SonarLint.CSharp/Helpers/ControlFlowGraph/ControlFlowGraphBuilder.cs#L659): Complete the task associated to this "TODO" comment. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Pull request merged**